src/functions.rs

```diff
@@ -2969,19 +3005,58 @@ pub fn search_for_recursion(
     }
 
     if !functions.is_empty() {
-        let mut error: Option<CompileError> = None;
+        let mut loops: Vec<Vec<String>> = Vec::new();
 
         for func in functions {
-            if let Some(function_info) = function_map.get(func) {
-                error = Some(add_or_create_compile_error(
-                    error,
-                    "Recursive Function call found",
-                    function_info.declaration_file,
-                    function_info.get_declaration_range().unwrap_or_default(),
-                    "These calls have been found in a recursive loop.  There may be more than one recursive loop.",
-                ));
+            loops.push(find_recursion_loop(
+                func,
+                function_map,
+                terminated_list,
+                &mut Vec::new(),
+            )?);
+        }
+
+        // Now we find the smallest loop and tell that to the user
+        let mut smallest: i32 = -1;
+        let mut smallest_loop: Vec<String> = Vec::new();
+        for current_loop in loops {
+            let size = current_loop.len() as i32;
+            if smallest == -1 || size < smallest {
+                smallest = size;
+                smallest_loop = current_loop;
             }
         }
+
+        let mut previous_function: String = String::new();
+        let mut error: Option<CompileError> = None;
+        for func in smallest_loop {
+            if let Some(function_info) = function_map.get(&func) {
+                // This is the start of the loop, so we want a slightly different message
+                if error.is_none() {
+                    error = Some(add_or_create_compile_error(
+                        error,
+                        "Recursive Function call found",
+                        function_info.declaration_file,
+                        function_info.get_declaration_range().unwrap_or_default(),
+                        "This function is the start of the loop.",
+                    ));
+                    previous_function = function_info.name.clone();
+                } else {
+                    error = Some(add_or_create_compile_error(
+                        error,
+                        "Recursive Function call found",
+                        function_info.declaration_file,
+                        function_info.get_declaration_range().unwrap_or_default(),
+                        &format!(
+                            "This function calls the previous function: {}.",
+                            previous_function
+                        ),
+                    ));
+                    previous_function = function_info.name.clone();
+                }
+            }
+        }
+
         // Unwrap is safe since we need to go through the loop above at least once
         return Err(CascadeErrors::from(error.unwrap()));
```

Comment:
I haven't looked at this yet, but I tried running this on my locally modified copy of refpolicy3 and it's panicing on this unwrap.

Reply:
Specifically, what I tested is that I merged this branch with the branch from this PR: https://github.com/dburgener/cascade/pull/147 and built Cascade.

Then I ran it against this refpolicy3 branch: https://github.com/dburgener/refpolicy3/tree/edits-for-recursion

The output after that is:

```
$ ../cascade/target/debug/casc policy/*
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/functions.rs:3061:46
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```